PIPS/pip-2.md

```diff
@@ -0,0 +1,115 @@
+---
+pip: 2
+title: Relayer Fees
+status: Draft
+author: John Wong (@jwheur)
+created: 2020-06-30
+---
+
+## Simple Summary
+The current LockProxy implementation does not have an in-built method for relayers to charge fees for their work.
+
+This PIP proposes a simple method to charge fees.
+
+## Motivation
+Network fees can be very expensive. For example, the total Ethereum network fees for some popular DApps can cost up to 500k USD over 30 days (source: https://ethgasstation.info/).
+
+## Specification
+
+Example fee deduction implementation for lock:
+```
+struct TxArgs {
+  bytes fromAssetHash;
+  bytes toAssetHash;
+  bytes toAddress;
+  uint256 amount;
+  uint256 feeAmount;
+  bytes feeReceiverAddr;
+}
+
+function lock(
+  address fromAssetHash,
+  uint64 toChainId,
+  bytes memory targetProxyHash,
+  bytes memory toAssetHash,
+  bytes memory toAddress,
+  uint256 amount,
+  bool deductFeeInLock,
+  uint256 feeAmount,
+  bytes memory feeReceiverAddr
+)
+  public
+{
+  bytes32 key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash);
+  require(registry[key] == true);
+
+  // Use SafeMath to ensure balances do not overflow
+  balances[key] = balances[key].add(amount);
+
+  // transfer tokens from user to LockProxy
+  require(_transferToContract(fromAssetHash, amount), "transfer asset from fromAddress to lock_proxy contract failed!");
+
+  TxArgs memory txArgs = TxArgs({
+      fromAssetHash: fromAssetHash,
+      toAssetHash: toAssetHash,
+      toAddress: toAddress,
+      amount: remainingAmount,
+      feeAmount: feeAmount,
+      feeReceiverAddr: feeReceiverAddr
+  });
+
+  uint256 remainingAmount = amount;
+  if (feeAmount > 0 && deductFeeInLock) {
+    // ensure that there is no overflow
+    remainingAmount = amount.sub(feeAmount);
```

Comment:
replace remainingAmount with txArgs.amount

Reply:
ah yes, will fix it